src/transformers/modeling_utils.py

```diff
@@ -2439,17 +2439,24 @@ def _init_added_embeddings_weights_with_mean(
         mean_embeddings = torch.mean(old_embeddings_weight, axis=0)
         old_centered_embeddings = old_embeddings_weight - mean_embeddings
         covariance = old_centered_embeddings.T @ old_centered_embeddings / old_num_tokens
-        if old_embedding_dim >= old_num_tokens:
-            # Covarince matrix must be positive definite. For edge cases, when `vocab_size` is
-            # smaller than `hidden_size`, covarince matrix won't be positive definite so we
-            # must add the eye matrix to the covarince matrix to convert it to be positive definite.
-            covariance = covariance + torch.eye(old_embedding_dim, device=old_embeddings.weight.device) * 1e-3
-        distribution = torch.distributions.multivariate_normal.MultivariateNormal(
-            mean_embeddings, covariance_matrix=1e-5 * covariance
+
+        # Check if the covariance is positive definite.
+        is_covariance_psd = bool(
+            (covariance == covariance.T).all() and (torch.linalg.eigvals(covariance).real >= 0).all()
         )
-        new_embeddings.weight.data[-1 * added_num_tokens :, :] = distribution.sample(
-            sample_shape=(added_num_tokens,)
-        ).to(old_embeddings.weight.dtype)
+        if is_covariance_psd:
+            # If covariances is positive definite, a distribution can be created. and we can sample new weights from it.
+            distribution = torch.distributions.multivariate_normal.MultivariateNormal(
+                mean_embeddings, covariance_matrix=1e-9 * covariance
```

Comment:
could have reduce the strictness of the test as well, not sure what's best?

Reply:
It was my first choice to do that but the test is pretty loose. with atol = 1e-3 and rtol = 1e-1.
The Recerrunt gemma tests failed with rtol=2.155319929122925 and atol=0.001309454208239913
Making the rtol higher than 2 is not the best choice IMO, so I decided to reduce the covariance, given that the most important thing is to initialize the added embeddings with the mean, the added noise from the covariance is not a big deal.

WDYT?

Reply:
yep good decision!